### DISCUSSION TITLE
Include `version_enforcer.h` in all headers in CGAL

### DISCUSSION THREAD

**Discussion by janetournois** [CGAL/cgal#7525] [Ideas]
@CGAL/geometryfactory what would you think about including the `version_enforcer.h` file to every header in CGAL, and adding the corresponding file to every release?
It would sometimes help find inconsistencies between the versions that are used, in particular in specific projects development.

Then, in projects outside the repo, use a local `version_enforcer.h`, to make sure the user is using the right version of CGAL.

**Comment by lrineau** [comment]
Why not include it in `<CGAL/config.h>`?

**Comment by MaelRL** [comment]
Can you please illustrate the first use case? The file uses a hard-coded version number that is checked against the CGAL version macros; if you put the file within CGAL, then the macro must have the version number, and then what is its purpose?

**Comment by janetournois** [comment]
@MaelRL In the "normal" setup it is useless.
When you start using copied/forked header files, it would be nice to be able to check at compile time whether the forked files (still) "belong to" the same version of CGAL as the version of CGAL read in `CGAL_DIR` (without introducing an extra-fork by including version_enforcer.h)

@lrineau it sounds good... I'm wondering if it covers all our fork-cases

**Reply by sloriot** [reply]
To do that you need to have a version macro in each file, including a header does not help.

**Reply by MaelRL** [reply]
If you're forking specific files, you'd have to tag them with something that lives entirely within the forked file (or in your patch repository) since you can't "trust" the rest of the CGAL headers.

You could define a version macro in your forked files and have the `verison_enforcer.h` in CGAL (`config.h` or somewhere else), but that's equivalent to putting a `version_enforcer.h` in the patch repo.

**Reply by sloriot** [reply]
Indeed you need a macro per forked file.

**Reply by janetournois** [reply]
My idea is
* add a straightforward `version_enforcer.h` in the CGAL repo, and include it everywhere (maybe through `config.h`)
* add a specialized `version_enforcer.h` in my fork, that is included BEFORE the one of CGAL, and pre-defines the acceptable CGAL version

**Reply by janetournois** [reply]
The aim is to avoid adding `#include <version_enforcer.h>` in each forked file (because the include is already there in the original CGAL file)

**Comment by lrineau** [comment]
## Jane's proposal

Let me clear what I have understood. The goal of Jane is to enforce the same version in two files:
  - `<CGAL/version.h>`, and
  - `<CGAL/version_enforcer.h>`.

By default, without any user-specific modified headers, both would come from a distribution of CGAL, and would of course agree.

With user-specific modified headers, a specific `CGAL/version_enforcer.h` would be shipped with the modified headers, and it will ensure that the version of the CGAL distribution is the one expected by the modified headers.

@janetournois: Is that a correct rephrasing of your idea, Jane ?


## Mael's proposal

Another proposal from Mael would be to encode the CGAL version in all CGAL headers, and make sure they can only be used in the right CGAL distribution.

**Reply by MaelRL** [reply]
No, this isn't what I'm saying. I was just trying to understand Jane's proposal (specifically the first part, where there is no forked files and where the file is useless then).

I don't have any strong objection to this mechanism to automatically include the version enforcer, but the version_enforcer.h (whichever its implementation) should become obsolete rather soon anyway since we will not have forked files living outside CGAL releases.

**Reply by janetournois** [reply]
that's right...

**Reply by lrineau** [reply]
The cost of adding a trivial header `version_enforcer.h` and 
```c++
#include CGAL/version_enforcer.h>
```
in `<CGAL/config.h>` is near zero, anyway. I propose we do that.

We can even add it in CGAL-5.6, because it is really harmless.

**Reply by lrineau** [reply]
@janetournois Please create the pull-request. We will merge it.

**Comment by janetournois** [comment]
@lrineau yes, exactly.

It seems that Mael's proposal is another solution to the same problem.

**Reply by lrineau** [reply]
The cost of having the CGAL version in all CGAL headers is high: for each new release, that would mean modifying more than 3400 headers. We can do it, but if we are sure of the benefit.

**Comment by janetournois** [comment]
Done in PR #7528
